### ISSUE TITLE
Crash since 4.7.X if i.imgur.com can't be accessed

### ISSUE THREAD

**Issue by rdmir3** [ImpactDevelopment/ImpactIssues#1460]
Setup:
- Default `Impact` folder with no mods turned on.
- 4.7.X without Optifine/Liteloader/MultiMC (4.6 not broken)

But also:
- i.imgur.com can't be accessed (DNS redirects it to a different domain, with an HTML page instead of an image file) 

Reproduce:
- Delete `Impact/assets`
- Run game
- When the assets folder is created, quit the game. (At the end of the step, some of the assets files should be of HTML format instead of image)
- Run game
- Crash

**If you can't find which commit caused it, send me a zip of all 4.7 betas' JSONs, so that I could try all of them to see which started it.** I'm willing to do my part of this bugfix.

I believe the first line of the crash report sums up the situation:

```
---- Minecraft Crash Report ----
// Would you like a cupcake?

Time: 7/24/19 11:50 AM
Description: Registering texture

java.lang.NullPointerException: Registering texture
	at impact.0l.a(Unknown Source)
	at cdh.a(SourceFile:84)
	at cdr.a(SourceFile:60)
	at impact.0r.0(Unknown Source)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1118)
	at impact.0r.0(Unknown Source)
	at impact.0r.0(Unknown Source)
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1321)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at impact.0r.0(Unknown Source)
	at impact.5o.0(Unknown Source)
	at impact.4z.0(Unknown Source)
	at bib.handler$init$zzl000(SourceFile:3097)
	at bib.aq(SourceFile:583)
	at bib.a(SourceFile:404)
	at net.minecraft.client.main.Main.main(SourceFile:123)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Client thread
Stacktrace:
	at impact.0l.a(Unknown Source)
	at cdh.a(SourceFile:84)

-- Resource location being registered --
Details:
	Resource location: minecraft:downloaded/2378311fabbccae299383fa84b1c8126b4c497ce
	Texture object class: impact.4k
Stacktrace:
	at cdr.a(SourceFile:60)
	at impact.0r.0(Unknown Source)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1118)
	at impact.0r.0(Unknown Source)
	at impact.0r.0(Unknown Source)
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1321)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at impact.0r.0(Unknown Source)
	at impact.5o.0(Unknown Source)
	at impact.4z.0(Unknown Source)
	at bib.handler$init$zzl000(SourceFile:3097)
	at bib.aq(SourceFile:583)

-- Initialization --
Details:
Stacktrace:
	at bib.a(SourceFile:404)
	at net.minecraft.client.main.Main.main(SourceFile:123)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)

-- System Details --
Details:
	Minecraft Version: 1.12.2
	Operating System: Mac OS X (x86_64) version 10.14.6
	Java Version: 1.8.0_74, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode), Oracle Corporation
	Memory: 122758688 bytes (117 MB) / 268435456 bytes (256 MB) up to 2147483648 bytes (2048 MB)
	JVM Flags: 7 total; -Xmx2G -XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32M
	IntCache: cache: 0, tcache: 0, allocated: 0, tallocated: 0
	Launched Version: 1.12.2-Impact_4.7.2
	LWJGL: 2.9.2
	OpenGL: AMD Radeon R9 M370X OpenGL Engine GL version 2.1 ATI-2.11.20, ATI Technologies Inc.
	GL Caps: Using GL 1.3 multitexturing.
Using GL 1.3 texture combiners.
Using framebuffer objects because ARB_framebuffer_object is supported and separate blending is supported.
Shaders are available because OpenGL 2.1 is supported.
VBOs are available because OpenGL 1.5 is supported.

	Using VBOs: Yes
	Is Modded: Very likely; Jar signature invalidated
	Type: Client (map_client.txt)
	Resource Packs: LiLiLi (incompatible)
	Current Language: English (US)
	Profiler Position: N/A (disabled)
	CPU: 8x Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
```

**Comment by LeafHacker** [comment]
Probably related to @leijurv's cool new random wallpaper feature

**Comment by StijnSimons** [comment]
Incorrect, thats 4.8. This has probably something to do with the [Resources](https://github.com/ImpactDevelopment/Resources)

**Comment by leijurv** [comment]
I don't think it is actually, I made sure that the client didn't crash when I turned my wifi off. 

If I'm reading this issue correctly, the crash is when the imgur url does return data, but it's an HTML page instead of a texture.

**Comment by StijnSimons** [comment]
Maybe an image that got removed?

**Comment by rdmir3** [comment]
> If I'm reading this issue correctly, the crash is when the imgur url does return data, but it's an HTML page instead of a texture.

Of course,

```
Description: Registering texture

java.lang.NullPointerException: Registering texture
```

You can try the following:
In the source code, replace one of the imgur links with `https://google.com`. Then:
- Delete `Impact/Assets`.
- Launch the client. Keep `Impact/Assets`.
- Launch the client for the second time.

> Maybe an image that got removed, which led to a 404 HTML page

`i.imgur.com` displays the 404 page as an image.

**Comment by LeafHacker** [comment]
I mean, we could probably just catch the exception and handle things from there as a last resort

**Comment by noobsplicer** [comment]
ty im not sure if there's a fix here (im dumb) but still

**Comment by EverythingIsTakenLOL** [comment]
Idk what #1460 is but all of the files in Impact/assets don't have a file extension their file type is just "File"

**Comment by rdmir3** [comment]
You can try renaming them to add `.png` extension to all of them and try to open them in a photo viewer or internet browser. At least one of them should show an error.

**Comment by StijnSimons** [comment]
`doobie#9482` said this issue still persists in 4.7.3.
[Stacktrace](https://pastebin.com/Q63t9xr0)

**Comment by StijnSimons** [comment]
His issue got resolved.
1. Stop minecraft
2. Delete the Impact folder (%appdata%\.minecraft\Impact)
3. Restart minecraft

**Comment by LeafHacker** [comment]
> His issue got resolved.
> 
> 1. Stop minecraft
> 2. Delete the Impact folder (%appdata%.minecraft\Impact)
> 3. Restart minecraft

That isn't a resolution, it's a workaround. This is a bug in the way we handle downloading texture files, we need to fix it in the source code.

**Comment by StijnSimons** [comment]
> > His issue got resolved.
> > 
> > 1. Stop minecraft
> > 2. Delete the Impact folder (%appdata%.minecraft\Impact)
> > 3. Restart minecraft
> 
> That isn't a resolution, it's a workaround. This is a bug in the way we handle downloading texture files, we need to fix it in the source code.

Its not a workaround, its an resolution in his case. His `i.imgur.com` domain got blocked from his / her network before trying it using a vpn.

**Comment by leijurv** [comment]
ok i mean it's just pedantics 😂 depends on which bug you're talking about, first is this issue where imgur being inaccessible writes corrupted data to disk in impact's texture assets, second is #1517 where having this corrupted data on disk crashes the client on load even if imgur is accessible. deleting impact doesn't really "resolve" either one, but it sorta works around the second one

**Comment by LeafHacker** [comment]
Love me some semantics... And pedantics too!

**Comment by xhBIROhx** [comment]
My baritone (1.13) can't open doors and trapdoors so the bot is stick. Someone help me pls or fix this bug. I also dowloaded the baritone for 1.14 and have the same problem. If someone fix it it woud be great.

**Comment by LeafHacker** [comment]
@xhBIROhx please don't post unrelated stuff on some random issue. File your own issue. If you're issue relates to baritone, it should be filed on baritone's issue tracker not impact's. Also you can't fix bugs in baritone unless you're willing to learn to code and then diagnose and fix the issue. Rather report it with enough detail that leijurv can reproduce and fix it.

**Comment by leijurv** [comment]
Effectively fixed in 4.8, thanks to #1652's fix and also switching to our own system for capes and assets. We will be able to retroactively fix this for 4.7.3, possibly, by changing referenced URLs away from imgur.com to our own domain.
